docs/rfcs/021-raii-runtime.md

```diff
@@ -0,0 +1,209 @@
+    - What should happen to operations running on the tenant when we change the state to this one?
+    - Is it better to change to `Terminating`?
+7. `Shutdown` - tenant is shut down, background operations are finished, and ready to recycle. It is the final state.
+    - Is it better to change to `Terminated`?
+
+Transitions:
+
+1. `Infant` -> `Loading`, `Downloading`.
+2. `Loading` -> `Active`, `Broken`.
+3. `Broken` is the final state.
+4. `Downloading` -> `Active`, `Broken`.
+5. `Active` -> `ShuttingDown`.
+6. `ShuttingDown` -> `Shutdown`.
+7. `Shutdown` is the final state.
+
+The transition graph has no cycles, which is a good property.
+
+Generally, we have 3 subsets of states:
+
+1. Paused - tries to load itself: `Loading`, `Downloading`.
+2. Alive - can operate: `Infant`, `Active`.
+3. Dead - cannot launch any operations, possibly already dropped: `Broken`, `ShuttingDown`, `Shutdown`.
+
+## `TenantAccessor`
+
+The first part of the chapter will be about `TenantAccessor` specifically. Later, we'll discuss `TimelineAccessor`.
+
+### Top-level interface
+
+We need a proper tree-like structure with a bare minimum of global states to use RAII, unlike carefully using `shutdown_tasks` as we do now.
+
+The idea is to have the following tenant API:
+
+```rust
+/// Gets the alive tenant from the in-memory data, then applies the specified
+/// function. Returns error if there is no tenant in memory or the tenant is not
+/// ready to launch the closure.
+pub fn with_tenant<F, O>(tenant_id: TenantId, func: F) -> anyhow::Result<O>
+where
+    F: FnOnce(TenantSyncRef) -> O {}
+
+/// Gets the alive tenant from the in-memory data, then applies the specified
+/// asynchronous function. Returns error if there is no tenant in memory or
+/// tenant is not ready to launch the closure.
+pub async fn with_tenant_async<F, T, O>(tenant_id: TenantId, func: F) -> anyhow::Result<O>
+where
+    F: FnOnce(TenantAsyncRef) -> T,
+    T: Future<Output = O> {}
+```
+
+So, _we won't have access to the tenant directly_. Instead, we only use transaction-inspired API with some closure that can access the tenant. These functions will be referred to as closures.
+
+- `TenantSyncRef` is `&Tenant`, but with a limited interface not to allow simple mistakes like waiting for something inside.
+- `TenantAsyncRef` is `Arc<Tenant>` only because of current Rust's limitations of async closures. Later, it will also be a `&Tenant`. It also has a limited interface to prevent some kinds of mistakes.
+
+_These closures **are not supposed to run any long-running code**, and should be checked on code reviews._
+
+To solve a problem, we'll have a global private mapping from `TenantId` to `TenantAccessor`, accessible only by `with_tenant`. `TenantAccessor` is a kind of guard similar to `Arc` in Rust but not the same. It allows running the closures with the tenant in scope but won't run them if it's not in the `Active` state, and this accessor is responsible for all cancellations.
```

Comment:
Would this mapping replace the `TENANTS` hash table we currently have  in tenant_mgr.rs?

Reply:
No. We'll still have mapping because we need some singleton to store operating tenants. It's possible to have a total RAII here, but we want the tenant to live until we decide to shut it down.